prototypes/ScheduledMergesTest.hs

```diff
@@ -294,20 +307,50 @@ completeM (MCompleted _ _ r)   = r
 completeM (MOngoing mt _ _ rs) = mergek mt (map getNonEmptyRun rs)
 
 instance Arbitrary T where
-  arbitrary = QC.frequency
-      [ (1, TCompleted <$> arbitrary)
-      , (1, TOngoing <$> arbitrary)
-      , (1, do
-          (incoming, tree) <- arbitrary
-             `QC.suchThat` (\(i, t) -> length i + length t > 0)
-          return (TPendingLevel incoming tree))
-      , (1, do
-          n <- QC.frequency
-            [ (3, pure 2)
-            , (1, QC.chooseInt (3, 8))
+  arbitrary = QC.sized $ \s -> do
+      n <- QC.chooseInt (1, max 1 s)
+      go n
+    where
+      -- n is the number of constructors of T + pre-existing runs in the tree
+      go n | n < 1 = error ("arbitrary T: n == " <> show n)
+      go n | n == 1 =
+          QC.frequency
+            [ (1, TCompleted <$> arbitrary)
+            , (1, TOngoing <$> arbitrary)
             ]
-          TPendingUnion <$> QC.vectorOf n (QC.scale (`div` n) arbitrary))
-      ]
+      go n =
+          QC.frequency
+            [ (1, do
+                -- pending level merge without child
+                preExisting <- QC.vector (n - 1)  -- 1 for constructor itself
+                return (TPendingLevel preExisting Nothing))
+            , (1, do
+                -- pending level merge with child
+                numPreExisting <- QC.chooseInt (0, min 20 (n - 2))
+                preExisting <- QC.vector numPreExisting
+                tree <- go (n - numPreExisting - 1)
+                return (TPendingLevel preExisting (Just tree)))
+            , (2, do
+                -- pending union merge
+                ns <- QC.shuffle =<< arbitraryPartition2 n
+                TPendingUnion <$> traverse go ns)
+            ]
+
+      -- Split into at least two smaller positive numbers. The input needs to be
+      -- greater than or equal to 2.
+      arbitraryPartition2 :: Int -> QC.Gen [Int]
+      arbitraryPartition2 n = assert (n >= 2) $ do
+          first <- QC.chooseInt (1, n-1)
+          (first :) <$> arbitraryPartition (n - first)
+
+      -- Split into smaller positive numbers.
+      arbitraryPartition :: Int -> QC.Gen [Int]
+      arbitraryPartition n
+            | n <  1 = return []
+            | n == 1 = return [1]
+            | otherwise = do
+              first <- QC.chooseInt (1, n)
+              (first :) <$> arbitraryPartition (n - first)
```

Comment:
These are not very fair partitions. Is that bad?

Reply:
I think it's generally good to have varying sizes of sub-trees and these get shuffled afterwards, so there is no correlation between size and order of the children. It also has the interesting effect that for overall size `n` we get `log n` children on average, each of which has average size `n / log n`. So increasing the size both leads to larger branching factor as well as deeper trees.